x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -326,8 +317,10 @@ protected PhysicalPlan rule(TopNExec topNExec, LocalPhysicalOptimizerContext ctx
             PhysicalPlan plan = topNExec;
             PhysicalPlan child = topNExec.child();
 
-            boolean canPushDownTopN = child instanceof EsQueryExec
-                || (child instanceof ExchangeExec exchangeExec && exchangeExec.child() instanceof EsQueryExec);
+            boolean canPushDownTopN = (child instanceof EsQueryExec esQueryExec && esQueryExec.canPushSorts())
+                || (child instanceof ExchangeExec exchangeExec
+                    && exchangeExec.child() instanceof EsQueryExec esQuery
+                    && esQuery.canPushSorts());
```

Comment:
Optional / nit: the check might be a bit more readable if extracting the instance check and canPushSorts() call into a small method?

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/108747/commits/3161d88feb7391595f7ced7b8a9a42760b483b98